app/src/main/java/com/example/goodnote/ui/viewModels/NoteViewModel.kt

```diff
@@ -1,30 +1,93 @@
 package com.example.goodnote.ui.viewModels
 
-import android.content.Context
+import android.util.Log
 import androidx.lifecycle.LiveData
 import androidx.lifecycle.MutableLiveData
 import androidx.lifecycle.ViewModel
 import androidx.lifecycle.viewModelScope
-import com.example.goodnote.database.repository.NoteRepo
 import com.example.goodnote.database.models.Note
-import com.example.goodnote.utils.Injectors
+import com.example.goodnote.database.repository.NoteRepo
+import com.example.goodnote.utils.dummyNotes
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
 import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
 // AndroidViewModel(application) - use this one when you need application context
-// when do you need application context? repo/network only? or?
-class NoteViewModel(private val context: Context) : ViewModel() {
+// when do you need application context? repo/network only
+class NoteViewModel(private val repository: NoteRepo) : ViewModel() {
 
-    // inject repository here via viewmodelfactory, so the views are completely decoupled from repos
+    private val _notes: MutableLiveData<List<Note>> = MutableLiveData(dummyNotes)
+    val notes: LiveData<List<Note>>
+        get() = _notes
 
-    private val repository: NoteRepo = Injectors.getNoteRepository(context)
 
-    // dummy lists
-   private lateinit var _dummyNotes: MutableLiveData<List<Note>>
-    val dummyNotes: LiveData<List<Note>>
-        get() = _dummyNotes
+    private var _repoNotes: MutableLiveData<List<Note>> = MutableLiveData()
+    val repoNotes: LiveData<List<Note>>
+        get() = _repoNotes
+
+    init {
+        getAllNotes()
+    }
 
     // all repo functions here
-    fun saveNote(note: Note) = viewModelScope.launch {
+
+    fun getAllNotes() {
+         viewModelScope.launch(Dispatchers.IO) {
+            val listIO = repository.getAllNotes()
+            withContext(Dispatchers.Main) {
+                 _repoNotes.postValue(listIO)
+            }
+        }
+    }
+
+    fun saveNote(note: Note) = viewModelScope.launch(Dispatchers.IO) {
+
+        // if title.isEmpty || title == null --> title = DEFAULT_TITLE
+
         repository.saveNote(note)
+        getAllNotes()
```

Comment:
Maybe it is not the best idea to load all notes again every time when a new note is saved :)

It is better practise to save note immediately in adapter when you add it to list, and then just call `saveNote` command on the repository. If the command is successful nothing should change since the note is already added to the list. If the `saveNote` command fails you should just remove the note from the list and notify the user that action has failed. Maybe you can even give them an option to retry this action.

Reply:
How would I get the autogenerated id for the new note?

Once the note in the list is clicked, a new activity opens along with an extra noteId. That noteId is used to load the note from the database.

Room does return a Long, so that would be the new note's Id, but then I'd still have to return that (new) ID to the note in list so that each note in the list can actually be clicked on

Reply:
You can create UIID id for the note in the view model. :) 
This id should be different from row id that is saved in the Notes database table.

Here you can see how UIID looks like: https://www.uuidgenerator.net/

This is how you can generate uiid in Kotlin: 
`val noteId = UUID.randomUUID().toString()`

Reply:
Should I then just remove the autogenerate id from the model/database completely? Or does it have some use?

Reply:
Also, now a new item is added to _repoNotes/_tags directly. 
However, it's unclear to me why I couldn't just turn the value to a mutable list, add a note/tag, then return it as a list, and set value to be that new list with an added item. It didn't work. What did work was basically doing the same thing, except it merged two immutable lists. How come?